Design.md

```diff
@@ -0,0 +1,241 @@
+    
+_**Note: "Super Admin" account has full Account Management access and can use "AssumeRoleBackbeat" to assume a role of another Account.**_
+
+## Tenant APIs
+<u>**Important Assumptions:**</u>
+* All the Tenant APIs on OSIS will use **Super Admin credentials** to manage on Vault.
+* A Vault Account is equivalent of OSE Tenant.
+
+### Create Tenant
+This API creates a Tenant on Vault.  
+1. Vault `create-account` api will be called using vaultclient. 
+    * Tenant Name is stored as Account Name  
+    * `cd_tenant_ids` list as part of the request to be stored as a property of the vault account.
+    * Account ID to be sent back as storage tenant ID to OSE
+1. `generate-account-access-key` for the account using vaultclient.
+1. Create a role with
+   1. Role name : `osis` (So role-arn can be generated in the format `arn:aws:iam::[account-id]:role/osis`
+   1. Trust policy
+   ```json
+    {
+        "Version": "2012-10-17",
+        "Statement": [{
+            "Effect": "Allow",
+            "Principal": {
+                "Service": "service-name"
+            },
+            "Action": "sts:AssumeRole"
+            }]
+    }
+   ```
+1. `attach-role-policy` with s3 and iam full access to this role
+1. `delete-access-key` for the account.
+
+### Create Tenant Example Activity Diagram
+![Create Tenant](Design_Files/OSE-OSIS-Create-Tenant.png)
+
+### Query Tenants
+This API will query tenants on Vault using a `filter` parameter.
+1. Vault `list-accounts` api will be called using vaultclient.
+1. Filter the result accounts using the `filter` parameter in the request
+
+### List Tenants
+This API will list tenants on Vault. Vault `list-accounts` api will be called using vaultclient.
+
+### Get Tenant
+This API will return the tenant on Vault with `tenantID`.
+1. Vault `get-account` api will be called using vaultclient with the provided `tenantID` as `AccountID`.
+
+### Head Tenant
+This API will check whether the tenant exists on Vault with `tenantID`.
+1. Vault `get-account` api will be called using vaultclient with the provided `tenantID` as `AccountID`.
+1. Return `true` or `false` if result is returned or not respectively.
+
+### Delete Tenant
+This API will delete the tenant on Vault. 
+
+**Approach 1:**
+1. Vault `delete-account` api will be called using vaultclient with the provided `tenantID` as `AccountID`.
+1. Return error if account is not empty.
+
+**Approach 2:** [suggested]
+1. Vault `delete-account` api will be called using vaultclient with the provided `tenantID` as `AccountID`.
+   1. If error due to account not empty, execute all the APIs below using Assumed Role for the account _(See User APIs)_
+        1. detach role policies 
+        1. delete roles
+        1. Remove users from groups
+        1. delete users
+        1. delete groups
+
+**The final decision is yet to be concluded.**
+
+### Update Tenant
+This API will enable or disable the tenant (HOLD the account? _Not implementing_)
```

Comment:
We do not implement it. Cool.
Question: Is it not going to be implemented at all or is it not going to be implemented for S3C-3938?
If its only for S3C-3938 then it needs to be mentioned.

Reply:
Actually this was left undecided in the initial discussion, as disabling a tenant should disable users to use object storage. Is there a way we can hold an account in the current Vault?

Reply:
Updated same in commit 59a5a88